internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+	ErrQOSHasShutdown   = errors.New("QOS has been shutdown")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+type serviceQOSHandler func(wrp.Message) (<-chan wrp.Message, <-chan struct{})
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	next serviceQOSHandler
+	// queue for wrp messages, ingested by serviceQOS
+	queue chan wrp.Message
+	// maxQueueSize is the allowable max size of the qos' priority queue, based on the sum of all queued wrp message's payload
+	maxQueueSize int
+
+	lock   sync.Mutex
+	ctx    context.Context
```

Comment:
By using a `queue` above, you don't need a context or cancel function.  Why?  The goroutine runs when the queue is around.  Close the queue and the routine is done.  Don't duplicate signals if at all possible.  It doesn't make things more reliable, it actually makes them less.

Reply:
that totally makes sense, why keep track of more things when you don't have to 🙂